pydantic_extra_types/types/phone_numbers.py

```diff
@@ -0,0 +1,51 @@
+from typing import Any, Callable, Generator
+from pydantic_core import PydanticCustomError, core_schema
+
+GeneratorCallableStr = Generator[Callable[[str | int], str], None, None]
+
+
+PhoneNumberError = PydanticCustomError("value_error", "value is not a valid phone number")
+
+
+class PhoneNumber(str):
+    """
+    An international phone number
+    """
+    default_region_code: str | None = None
+    format: str = "RFC3966"
+    min_length: int = 7
+    max_length: int = 64
+
+
+    @classmethod
+    def __get_pydantic_core_schema__(cls, **_kwargs: Any) -> core_schema.AfterValidatorFunctionSchema:
+        return core_schema.general_after_validator_function(
+            cls.validate,
+            core_schema.str_schema(
+                min_length=cls.min_length, max_length=cls.max_length
+            ),
+        )
+
+    @classmethod
+    def validate(cls, phone_number: str, _: core_schema.ValidationInfo) -> None:
+        import phonenumbers
+
+        try:
+            parsed_number = phonenumbers.parse(phone_number, cls.default_region_code)
+        except phonenumbers.phonenumberutil.NumberParseException as exc:
+            raise PhoneNumberError from exc
+        if not phonenumbers.is_valid_number(parsed_number):
+            raise PhoneNumberError
+
+
+
+        return phonenumbers.format_number(
+            parsed_number, getattr(phonenumbers.PhoneNumberFormat, cls.format)
+        )
+
+
+class USPhoneNumber(PhoneNumber):
```

Comment:
I'm not sure what other countries make sense to include. I like including this because it both serves as an example and as a common case for US developers.

Reply:
Looks good to me i test it locally 👍🏻 !

Here is my sense, defines a class called `PhoneNumberRegion` and a function called `_phone_number_region()` that returns a list of `PhoneNumberRegion` objects.

```py
@dataclass
class PhoneNumberRegion:
    region_code: str


@lru_cache()
def _phone_number_region() -> List[PhoneNumberRegion]:
    return [
        PhoneNumberRegion(
            'US',
        ),
```

what you think @JamesHutchison

Reply:
Would you be able to clarify the context of where this would be? I'm not quite understanding what the object of your suggestion is.

I updated the code to pull the valid regions and formats via a metaclass, as I found that nontrivial to look-up. I do see an opportunity to create `PhoneNumberRegion` and `PhoneNumberFormat` types but I'm also ok with just using strings. If there's existing patterns to follow, that might help with the direction.